### PR TITLE
Grants exp at unit kill

### PR DESCRIPTION
**Pull request by iwek7**
This solves issue when fighting and exp systems were not connected, now exp is granted to attacker at kill, formula for amount of exp is `killedUnitLevel * 8`. 

Also solves issue where unit level was not incremented on levelup.

Also solves issue where np was thrown when attacking unit with attack type that defending unit did not have.

Also solves bug where units with 0hp were not killed.

### CODE REVIEW

File: source/unit/Attack.gd [hunk 1]
```diff
@@ -11,7 +11,7 @@ export var icon : StreamTexture = null
 onready var specials := get_children()
 
 func execute_before_turn(user, target):
-	if not specials.empty():
+	if not specials == null and not specials.empty():
```
Comment (Byteron) [inline source/unit/Attack.gd]: `if not special == null and` equals to `if specials and` I think that's much easier to understand.
same goes for lines 24 and 34.

**Comment by Byteron** [comment]
can be merged if the comment above is resolved.

File: source/unit/Attack.gd (hunk 1 above)
Comment (iwek7) [reply source/unit/Attack.gd]: fixed, thanks

**Comment by GregoryLundberg** [comment]
This looks to me like several commits rolled into one.

* Fix a couple of whitespace issues, including a missing end-of-line and end-of-file.
* Guard against a null reference (the specials stuff)
* A gratuitous name change (unit->attacker) which hides the actual change
* Removing initialization for an unused variable (???)

It would be best to break these into individual subject commits so we can more accurately gage what you're actually doing and more easily ignore the noise.

**Comment by iwek7** [comment]
I see that there are some logic changes that I did not bother to put in separate commit but I should, I did it mostly because fighting is now Having some issues and there were some obvious things to repair when I learned the code. However moving small related refactor and formatting changes to separate commits seems to be too much, and frankly I don’t want to do it. I will split the commits by logic fixes tho, thanks for feedback!

File: source/unit/Unit.gd [hunk 2]
```diff
@@ -143,6 +139,7 @@ func _set_experience_current(value: int) -> void:
 func _amla() -> void:
 	type.health += 3
 	type.experience *= 1.2
+	type.level += 1
```
Comment (Byteron) [inline source/unit/Unit.gd]: I am pretty sure the level does not increase on amla, only XP and Health. The Level is more representative for the advancement level of the unit.

**Comment by Byteron** [comment]
Also, for each combat, a unit gains XP equal to the level of it's opponent if it does not die.
is that in already, or are XP only given on death?

**Comment by iwek7** [comment]
As for my knowledge in current master no xp is granted to any unit ever, however fighting code is complicated and perhaps I did not understand something (however just play game to see that it is the case). My code awards xp only when opponent is killed in battle.

File: source/unit/Unit.gd (hunk 2 above)
Comment (iwek7) [reply source/unit/Unit.gd]: ye, I removed it

**Comment by iwek7** [comment]
I also split code into 3 commits according to @GregoryLundberg wish

**Pull request merged**